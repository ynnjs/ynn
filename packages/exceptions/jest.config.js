/******************************************************************
 * Copyright (C) 2021 LvChengbin
 *
 * File: exceptions/jest.config.js
 * Author: LvChengbin<lvchengbin59@gmail.com>
 * Time: 05/03/2021
 * Description:
 ******************************************************************/

module.exports = {
    ...require( '../../jest.config.js' ),
    rootDir : __dirname,
    name : '@ynn/exceptions',
    displayName : '@ynn/exceptions'
};
