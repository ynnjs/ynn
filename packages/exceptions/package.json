{
    "name": "@ynn/exceptions",
    "version": "0.0.0",
    "author": "LvChengbin <lvchengbin59@gmail.com>",
    "license": "MIT",
    "scripts": {
        "clean": "rm -rf lib",
        "build": "yarn clean && tsc -b -v",
        "test": "jest",
        "lint": "eslint {src,test}/**/*.ts --fix --cache",
        "postpack": "run clean",
        "prepack": "run build"
    },
    "engines": {
        "node": ">14.15.0"
    },
    "main": "lib/index.js",
    "files": [
        "lib"
    ],
    "devDependencies": {
        "@types/statuses": "^2.0.0",
        "@ynn/jest-extended": "workspace:packages/jest-extended",
        "eslint": "^7.28.0",
        "eslint_d": "^10.1.3",
        "jest": "^27.0.4",
        "jest-extended": "^0.11.5",
        "ts-jest": "^27.0.3",
        "typescript": "^4.3.2"
    },
    "dependencies": {
        "@ynn/utility-types": "workspace:packages/utility-types",
        "statuses": "^2.0.1"
    }
}
