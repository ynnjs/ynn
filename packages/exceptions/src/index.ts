/******************************************************************
 * Copyright (C) 2021 LvChengbin
 *
 * File: exceptions/index.ts
 * Author: LvChengbin<lvchengbin59@gmail.com>
 * Time: 05/01/2021
 * Description:
 ******************************************************************/

export * from './http-exception';
