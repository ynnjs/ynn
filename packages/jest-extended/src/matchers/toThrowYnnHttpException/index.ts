/******************************************************************
 * Copyright (C) 2021 LvChengbin
 *
 * File: toThrowYnnHttpException/index.ts
 * Author: LvChengbin<lvchengbin59@gmail.com>
 * Time: 04/25/2021
 * Description:
 ******************************************************************/

import { CustomMatcherResult } from 'jest';
import { matcherHint, printExpected, printReceived } from 'jest-matcher-utils';
import { VariadicFunction } from '@ynn/utility-types';
import { HttpException, HttpExceptionResponse } from '@ynn/http-exception';
import predicate from './predicate';

const passMessage = ( received: unknown, expected: unknown ): ( () => string ) => (): string => {
    return matcherHint( '.toThrowYnnHttpException', 'function', 'type' ) +
        '\n\n' +
        'Expected not to throw:\n' +
        `  ${printExpected( expected?.toJSON?.() ?? expected )}\n` +
        'Thrown:\n' +
        `  ${printReceived( received?.toJSON?.() ?? received )}\n`;
};

const failMessage = ( received: unknown, expected: unknown ): ( () => string ) => (): string => {
    return matcherHint( '.not.toThrowYnnHttpException', 'function', 'type' ) +
        '\n\n' +
        'Expected to throw:\n' +
        `  ${printExpected( expected?.toJSON?.() ?? expected )}\n` +
        'Thrown:\n' +
        `  ${printReceived( received?.toJSON?.() ?? received )}\n`;
};

/**
 *
 * @example
 * ```ts
 * expect( () => { throw new HttpException( 400 ) } ).toThrowYnnHttpException()
 * ```
 *
 * @param callback - the callback function which will throw exception
 * @param response - status, message, response or HttpException instance
 * @param message - error message of HttpException instance.
 */
export default function toThrowYnnHttpException(
    callback: VariadicFunction,
    response?: number | string | HttpExceptionResponse | HttpException,
    error?: string
): CustomMatcherResult {

    console.log( 'xxxxxxxxxxxxxxxxxxxxxxxxxxxxx', callback );

    if( !callback || typeof callback !== 'function' ) {
        return {
            pass : false,
            message : (): string => `Recevied value must be a function but instead "${callback}" was found`
        };
    }

    let exception: any; // eslint-disable-line @typescript-eslint/no-explicit-any

    try { callback() } catch( e: unknown ) { exception = e }

    if( !exception ) {
        return {
            pass : false,
            message : (): string => 'Expect the fucntion to throw an HttpException.\n But it didn\'t throw anything'
        };
    }

    if( !( exception instanceof HttpException ) ) {
        return {
            pass : false,
            // message : failMessage( error, HttpException )
            message : (): string => `Expect the function to throw an HttpException.\n But ${exception} was thrown`
        };
    }

    /**
     * any HttpException instance will pass the checking if the response argument is empty
     */
    if( !response ) {
        return {
            pass : true,
            message : passMessage( exception, HttpException )
        };
    }

    if( typeof response === 'number' ) {
        if( error === undefined ) {
            if( exception.status === response ) {
                return {
                    pass : true,
                    message : passMessage( exception, { status : response } )
                };
            }

            return {
                pass : false,
                message : failMessage( exception, { status : response } )
            };
        }
        if( exception.status === response && exception.error === error ) {
            return {
                pass : true,
                message : passMessage( exception, { status : response, error } )
            };
        }

        return {
            pass : false,
            message : failMessage( exception, { status : response, error } )
        };
    }

    if( typeof response === 'string' ) {
        if( exception.error === response ) {
            return {
                pass : true,
                message : passMessage( exception, { error : response } )
            };
        }
        return {
            pass : false,
            message : failMessage( exception, { error : response } )
        };
    }

    let httpException: HttpException;

    if( response instanceof HttpException ) {
        httpException = response;
    } else {
        httpException = new HttpException( response, error );
    }

    if( !predicate( exception, response, error ) ) {
        return {
            pass : false,
            message : failMessage( exception, httpException )
        };
    }

    return {
        pass : true,
        message : passMessage( exception, new HttpException( response, error ) )
    };
}
