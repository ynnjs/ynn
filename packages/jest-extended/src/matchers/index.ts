/******************************************************************
 * Copyright (C) 2021 LvChengbin
 *
 * File: matchers/index.ts
 * Author: LvChengbin<lvchengbin59@gmail.com>
 * Time: 04/25/2021
 * Description:
 ******************************************************************/

import toThrowYnnHttpException from './toThrowYnnHttpException';

export default {
    toThrowYnnHttpException
};
