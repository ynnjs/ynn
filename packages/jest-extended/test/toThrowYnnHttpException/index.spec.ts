/******************************************************************
 * Copyright (C) 2021 LvChengbin
 *
 * File: toThrowYnnHttpException/index.spec.ts
 * Author: LvChengbin<lvchengbin59@gmail.com>
 * Time: 04/26/2021
 * Description:
 ******************************************************************/

import { HttpException } from '@ynn/exceptions';
import matchers from '../../src/matchers';

expect.extend( matchers );

describe( '.toThrowYnnHttpException', () => {
    it( 'passes when thrown an arbitrary HttpException', () => {
        expect( () => { throw new HttpException( 400 ) } ).toThrowYnnHttpException();
        expect( () => { throw new HttpException( 400, 'Something Error' ) } ).toThrowYnnHttpException();
        expect( () => { throw new HttpException( { status : 400, message : [ 'Something Error' ] } ) } ).toThrowYnnHttpException();
    } );

    it( 'passes when thrown a HttpException matched the expect HttpException instance', () => {
        expect( () => { throw new HttpException( 400 ) } ).toThrowYnnHttpException( new HttpException( 400 ) );
        expect( () => { throw new HttpException( 400, 'Something Error' ) } ).toThrowYnnHttpException( new HttpException( 400, 'Something Error' ) );
        const response = { status : 400, message : [ 'Something Error' ] };
        expect( () => { throw new HttpException( response ) } ).toThrowYnnHttpException( new HttpException( response ) );
    } );

    it( 'passes when thrown a HttpException and matched the given status', () => {
        expect( () => { throw new HttpException( 400 ) } ).toThrowYnnHttpException( 400 );
        expect( () => { throw new HttpException( 400, 'Something Error' ) } ).toThrowYnnHttpException( 400 );
        expect( () => { throw new HttpException( 400, 'Something Error' ) } ).toThrowYnnHttpException( 400 );
        expect( () => { throw new HttpException( { status : 400, message : [ 'Something Error' ] } ) } ).toThrowYnnHttpException( 400 );
    } );

    it( 'passes when thrown a HttpException and matched the given error message', () => {
        expect( () => { throw new HttpException( 400 ) } ).toThrowYnnHttpException( 'Bad Request' );
        expect( () => { throw new HttpException( 400, 'Something Error' ) } ).toThrowYnnHttpException( 'Something Error' );
        expect( () => { throw new HttpException( { status : 400, error : 'Something Error' } ) } ).toThrowYnnHttpException( 'Something Error' );
    } );

    it( 'passes when thrown a HttpException and matched both the given status and error message', () => {
        expect( () => { throw new HttpException( 400 ) } ).toThrowYnnHttpException( 400, 'Bad Request' );
        expect( () => { throw new HttpException( 400, 'Something Error' ) } ).toThrowYnnHttpException( 400, 'Something Error' );
        expect( () => { throw new HttpException( { status : 400, error : 'Something Error' } ) } ).toThrowYnnHttpException( 400, 'Something Error' );
    } );

    it( 'fails when not thrown any exception', () => {
        expect( () => {
            expect( () => {} ).toThrowYnnHttpException();
        } ).toThrowErrorMatchingSnapshot();
    } );

    it( 'fails when not thrown HttpException', () => {
        expect( () => {
            expect( () => { throw new Error() } ).toThrowYnnHttpException();
        } ).toThrowErrorMatchingSnapshot();
    } );

    it( 'fails when thrown a HttpException but not matched the given HttpException instance', () => {
        expect( () => {
            expect( () => { throw new HttpException( 401 ) } ).toThrowYnnHttpException( new HttpException( 400 ) );
        } ).toThrowErrorMatchingSnapshot();
        expect( () => {
            expect( () => { throw new HttpException( 400, 'Something Error' ) } ).toThrowYnnHttpException( new HttpException( 400 ) );
        } ).toThrowErrorMatchingSnapshot();
    } );

    it( 'fails when thrown a HttpException and not matched the given status', () => {
        expect( () => {
            expect( () => { throw new HttpException( 401 ) } ).toThrowYnnHttpException( 400 );
        } ).toThrowErrorMatchingSnapshot();

        expect( () => {
            expect( () => { throw new HttpException( 401, 'something error' ) } ).toThrowYnnHttpException( 400 );
        } ).toThrowErrorMatchingSnapshot();

        expect( () => {
            expect( () => { throw new HttpException( { status : 401, error : 'something error' } ) } ).toThrowYnnHttpException( 400 );
        } ).toThrowErrorMatchingSnapshot();
    } );

    it( 'fails when thrown a HttpException and not matched the given error message', () => {
        expect( () => {
            expect( () => { throw new HttpException( 400 ) } ).toThrowYnnHttpException( 'Something Error' );
        } ).toThrowErrorMatchingSnapshot();

        expect( () => {
            expect( () => { throw new HttpException( 400, 'something error' ) } ).toThrowYnnHttpException( 400, 'Bad Request' );
        } ).toThrowErrorMatchingSnapshot();

        expect( () => {
            expect( () => { throw new HttpException( { status : 401, error : 'something error' } ) } ).toThrowYnnHttpException( 400, 'Bad Request' );
        } ).toThrowErrorMatchingSnapshot();
    } );
} );

describe( '.rejects.toThrowYnnHttpException', () => {
    it( 'passes when the promise object rejected with an arbitrary HttpException instance', () => {
        expect( Promise.reject( new HttpException( 400 ) ) ).rejects.toThrowYnnHttpException();
    } );

    it( 'passes when the thrown HttpException matched the expected one', () => {
        const args = { status : 400, message : [ 'Something Error' ] };
        expect( Promise.reject( new HttpException( args ) ) ).rejects.toThrowYnnHttpException( new HttpException( args ) );
    } );

    it( 'passes when thrown a HttpException and matched the HttpException created with given object', () => {
        const args = { status : 400, message : [ 'Something Error' ] };
        expect( Promise.reject( new HttpException( args ) ) ).rejects.toThrowYnnHttpException( args );
    } );

    it( 'fails when the promise object not rejected with a HttpException instance', () => {
        expect( () => {
            return expect( Promise.reject( new Error() ) ).rejects.toThrowYnnHttpException();
        } ).rejects.toThrowErrorMatchingSnapshot();
    } );

    it( 'fails when the thrown HttpException did not match the expected one', () => {
        const args = { status : 400, message : [ 'Something Error' ] };
        expect( () => {
            return expect( Promise.reject( new HttpException( args ) ) ).rejects.toThrowYnnHttpException( new HttpException( 401 ) );
        } ).rejects.toThrowErrorMatchingSnapshot();
    } );
} );

describe( '.not.toThrowYnnHttpException', () => {
    it( 'passes when not thrown any exception', () => {
        expect( () => {} ).not.toThrowYnnHttpException();
    } );

    it( 'passes when not thrown a HttpException', () => {
        expect( () => { throw new Error() } ).not.toThrowYnnHttpException();
    } );

    it( 'passes when thrown a HttpException but not matched the given status', () => {
        expect( () => { throw new HttpException( 400 ) } ).not.toThrowYnnHttpException( 401 );
    } );

    it( 'passes when thrown a HttpException but not matched the given error message', () => {
        expect( () => { throw new HttpException( 400 ) } ).not.toThrowYnnHttpException( 'Something Error' );
    } );

    it( 'passes when thrown a HttpException but not matched the given response', () => {
        expect( () => { throw new HttpException( { status : 400, message : [ 'Bad Request' ] } ) } ).not.toThrowYnnHttpException( new HttpException( {
            status : 400,
            message : [ 'Something Error' ]
        } ) );
    } );

} );

describe( '.rejects.not.toThrowYnnHttpException', () => {
    it( 'passes when the promise not rejected with a HttpException instance', () => {
        expect( Promise.reject( new Error ) ).rejects.not.toThrowYnnHttpException();
    } );

    it( 'passes when the thrown HttpException instance did not match the expected one', () => {
        expect( Promise.reject( new HttpException( 400 ) ) ).rejects.not.toThrowYnnHttpException( new HttpException( 401 ) );
    } );

    it( 'failes when the promise rejected with a HttpException instance', () => {
        expect( () => {
            return expect( Promise.reject( new HttpException( 400 ) ) ).rejects.not.toThrowYnnHttpException();
        } ).rejects.toThrowErrorMatchingSnapshot();
    } );

    it( 'fails when the thrown HttpException matched the expected one', () => {
        expect( () => {
            return expect( Promise.reject( new HttpException( 400 ) ) ).rejects.not.toThrowYnnHttpException( new HttpException( 400 ) );
        } ).rejects.toThrowErrorMatchingSnapshot();
    } );
} );
