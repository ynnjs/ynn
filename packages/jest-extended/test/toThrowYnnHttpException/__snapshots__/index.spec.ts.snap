// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`.rejects.not.toThrowYnnHttpException failes when the promise rejected with a HttpException instance 1`] = `
"[2mexpect([22m[31mfunction[39m[2m).toThrowYnnHttpException([22m[32mtype[39m[2m)[22m

Expected not to throw:
  [32m[Function HttpException][39m
Thrown:
  [31m{\\"error\\": \\"Bad Request\\", \\"response\\": 400, \\"status\\": 400}[39m
"
`;

exports[`.rejects.not.toThrowYnnHttpException fails when the thrown HttpException matched the expected one 1`] = `
"[2mexpect([22m[31mfunction[39m[2m).toThrowYnnHttpException([22m[32mtype[39m[2m)[22m

Expected not to throw:
  [32m{\\"error\\": \\"Bad Request\\", \\"response\\": [Error], \\"status\\": 400}[39m
Thrown:
  [31m{\\"error\\": \\"Bad Request\\", \\"response\\": 400, \\"status\\": 400}[39m
"
`;

exports[`.rejects.toThrowYnnHttpException fails when the promise object not rejected with a HttpException instance 1`] = `
"Expect the function to throw an HttpException.
 But Error was thrown"
`;

exports[`.rejects.toThrowYnnHttpException fails when the thrown HttpException did not match the expected one 1`] = `
"[2mexpect([22m[31mfunction[39m[2m).not.toThrowYnnHttpException([22m[32mtype[39m[2m)[22m

Expected to throw:
  [32m{\\"error\\": \\"Unauthorized\\", \\"response\\": 401, \\"status\\": 401}[39m
Thrown:
  [31m{\\"error\\": \\"Bad Request\\", \\"response\\": {\\"message\\": [\\"Something Error\\"], \\"status\\": 400}, \\"status\\": 400}[39m
"
`;

exports[`.toThrowYnnHttpException fails when not thrown HttpException 1`] = `
"Expect the function to throw an HttpException.
 But Error was thrown"
`;

exports[`.toThrowYnnHttpException fails when not thrown any exception 1`] = `
"Expect the fucntion to throw an HttpException.
 But it didn't throw anything"
`;

exports[`.toThrowYnnHttpException fails when thrown a HttpException and not matched the given error message 1`] = `
"[2mexpect([22m[31mfunction[39m[2m).not.toThrowYnnHttpException([22m[32mtype[39m[2m)[22m

Expected to throw:
  [32m{\\"error\\": \\"Something Error\\"}[39m
Thrown:
  [31m{\\"error\\": \\"Bad Request\\", \\"response\\": 400, \\"status\\": 400}[39m
"
`;

exports[`.toThrowYnnHttpException fails when thrown a HttpException and not matched the given error message 2`] = `
"[2mexpect([22m[31mfunction[39m[2m).not.toThrowYnnHttpException([22m[32mtype[39m[2m)[22m

Expected to throw:
  [32m{\\"error\\": \\"Bad Request\\", \\"status\\": 400}[39m
Thrown:
  [31m{\\"error\\": \\"something error\\", \\"response\\": 400, \\"status\\": 400}[39m
"
`;

exports[`.toThrowYnnHttpException fails when thrown a HttpException and not matched the given error message 3`] = `
"[2mexpect([22m[31mfunction[39m[2m).not.toThrowYnnHttpException([22m[32mtype[39m[2m)[22m

Expected to throw:
  [32m{\\"error\\": \\"Bad Request\\", \\"status\\": 400}[39m
Thrown:
  [31m{\\"error\\": \\"something error\\", \\"response\\": {\\"error\\": \\"something error\\", \\"status\\": 401}, \\"status\\": 401}[39m
"
`;

exports[`.toThrowYnnHttpException fails when thrown a HttpException and not matched the given status 1`] = `
"[2mexpect([22m[31mfunction[39m[2m).not.toThrowYnnHttpException([22m[32mtype[39m[2m)[22m

Expected to throw:
  [32m{\\"status\\": 400}[39m
Thrown:
  [31m{\\"error\\": \\"Unauthorized\\", \\"response\\": 401, \\"status\\": 401}[39m
"
`;

exports[`.toThrowYnnHttpException fails when thrown a HttpException and not matched the given status 2`] = `
"[2mexpect([22m[31mfunction[39m[2m).not.toThrowYnnHttpException([22m[32mtype[39m[2m)[22m

Expected to throw:
  [32m{\\"status\\": 400}[39m
Thrown:
  [31m{\\"error\\": \\"something error\\", \\"response\\": 401, \\"status\\": 401}[39m
"
`;

exports[`.toThrowYnnHttpException fails when thrown a HttpException and not matched the given status 3`] = `
"[2mexpect([22m[31mfunction[39m[2m).not.toThrowYnnHttpException([22m[32mtype[39m[2m)[22m

Expected to throw:
  [32m{\\"status\\": 400}[39m
Thrown:
  [31m{\\"error\\": \\"something error\\", \\"response\\": {\\"error\\": \\"something error\\", \\"status\\": 401}, \\"status\\": 401}[39m
"
`;

exports[`.toThrowYnnHttpException fails when thrown a HttpException but not matched the given HttpException instance 1`] = `
"[2mexpect([22m[31mfunction[39m[2m).not.toThrowYnnHttpException([22m[32mtype[39m[2m)[22m

Expected to throw:
  [32m{\\"error\\": \\"Bad Request\\", \\"response\\": 400, \\"status\\": 400}[39m
Thrown:
  [31m{\\"error\\": \\"Unauthorized\\", \\"response\\": 401, \\"status\\": 401}[39m
"
`;

exports[`.toThrowYnnHttpException fails when thrown a HttpException but not matched the given HttpException instance 2`] = `
"[2mexpect([22m[31mfunction[39m[2m).not.toThrowYnnHttpException([22m[32mtype[39m[2m)[22m

Expected to throw:
  [32m{\\"error\\": \\"Bad Request\\", \\"response\\": 400, \\"status\\": 400}[39m
Thrown:
  [31m{\\"error\\": \\"Something Error\\", \\"response\\": 400, \\"status\\": 400}[39m
"
`;
